File: prac4lab/8.f90
module q8
real,parameter::pi=acos(-1.0)
integer::n,i,j,k
real::ans, prod,m
real,dimension(:),allocatable::x
real,dimension(:,:),allocatable::q

end module q8

program sx
use q8

read*,n,m

allocate(x(0:n), q(0:n,0:n))

do i =0,n
x(i)=cos((2*i+1)*pi/(2*n+2))
q(i,0)= sin(pi*x(i))
enddo

call dif()

do i=0,n
print*, (q(i,j), j=0,n)
enddo

ans = q(0,0)

do i=1,n
call prdct(i,m)
ans = ans + q(0,i)*prod
enddo

print*,ans


end program sx

subroutine dif()
use q8
integer::r,c

do c=1,n
do r=0,n-c
q(r,c)=(q(r+1,c-1)-q(r,c-1))/(x(r+c)-x(r))
enddo
enddo



end subroutine dif

subroutine prdct(a,b)
use q8
integer::a
real::b

prod =1

do k=0,a-1
prod = prod* (b-x(k))
enddo


end subroutine prdct

File: prac4lab/8.cpp
#include <bits/stdc++.h>
using namespace std;

/* name of program */

 vector<double> x; 
 vector<vector<double>> q;

double product(int a, double b)
{
    double prod = 1;

    for (int k = 0; k < a; k++)
    {
        prod *= (b - x[k]);
    }
    return prod;
}

int main()
{
    int n;
    double m;
    
    cin >> n>>m ;
    // double x[n+1],q[n+1][n+1];



    for (int i = 0; i <= n; i++)
    {
        x.push_back(cos((2 * i + 1) * M_PI / (2 * n + 2)));
        q.push_back({sin(M_PI * x[i])});
    }

    for (int c = 1; c <= n; c++)
    {
        for (int r = 0; r <= n - c; r++)
        {
            q[r].push_back ((q[r + 1][c - 1] - q[r][c - 1]) / (x[r + c] - x[r]));
        }
    }

    for (int i=0;i<=n;i++){
        for(auto x : q[i]){
            cout<<x<<" ";
        }
        cout<<endl;
    }

    double ans = q[0][0];

    for (int i = 1; i <= n; i++)
    {
        ans += q[0][i] * product(i, m);
    }

    cout << ans << endl;

    return 0;
}